File: proper-acronym-2/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

/* Authored by Kay Akashi */

void solve() {
    ll N; cin >> N;
    string S = " ";
    for (ll i = 0; i < N; i++) {
        string s; cin >> s;
        S += s.at(0);
    }
    string T; cin >> T;
    T = " " + T;

    ll Ls = S.size();
    ll Lt = T.size();

    vector<vector<ll>> dp = {};
    for (ll i = 0; i < Ls; i++) {
        vector<ll> cand = {};
        for (ll j = 0; j < Lt; j++) {
            cand.push_back(0);
        }
        dp.push_back(cand);
    }
    for (ll i = 1; i < Ls; i++) {
        for (ll j = 1; j < Lt; j++) {
            if (S.at(i) == T.at(j)) {
                dp.at(i).at(j) = max({dp.at(i - 1).at(j), dp.at(i).at(j - 1), dp.at(i - 1).at(j - 1) + 1});
            }
            else {
                dp.at(i).at(j) = max({dp.at(i - 1).at(j), dp.at(i).at(j - 1)});
            }
        }
    }
    
    if (Lt - 1 == dp.at(Ls - 1).at(Lt - 1)) {
        cout << "Yes" << endl;
    }
    else {
        cout << "No" << endl;
    }
}

int main() {
    solve();
}

File: proper-acronym-2/solution.py
def main():
	N = int(input())
	li = list(map(str, input().split()))
	S = " "
	for el in li:
		S += el[0]
	T = " " + str(input())

	Ls = len(S)
	Lt = len(T)

	dp = []
	for i in range(Ls):
		cand = [0 for j in range(Lt)]
		dp.append(cand)

	for i in range(1, Ls):
		for j in range(1, Lt):
			if S[i] == T[j]:
				dp[i][j] = max([dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + 1])
			else:
				dp[i][j] = max([dp[i - 1][j], dp[i][j - 1]])

	if dp[Ls - 1][Lt - 1] == Lt - 1:
		print("Yes")
	else:
		print("No")

if __name__ == '__main__':
	main()